odoo/addons/base/tests/test_ir_model.py

```diff
@@ -168,3 +169,112 @@ def test_create_external_id_with_space(self):
+
+class TestIrModel(SavepointCase):
+    
+    @classmethod
+    def setUpClass(cls):
+        res = super().setUpClass()
+        cls.registry.enter_test_mode(cls.cr)
+        cls.test_stage_model = cls.env['ir.model'].create({
+            'name': 'Banana Ripeness',
+            'model': 'x_banana_ripeness',
+            'field_id': [
+                (0, 0, {'name': 'x_name', 'ttype': 'char', 'field_description': 'Name'}),
+            ]
+        })
+        cls.ripeness_green = cls.env['x_banana_ripeness'].name_create('Green')[0]
+        cls.ripeness_okay = cls.env['x_banana_ripeness'].name_create('Okay, I guess?')[0]
+        cls.ripeness_gone = cls.env['x_banana_ripeness'].name_create('Walked away on its own')[0]
+        cls.test_model = cls.env['ir.model'].create({
+            'name': 'Bananas',
+            'model': 'x_bananas',
+            'field_id': [
+                (0, 0, {'name': 'x_name', 'ttype': 'char', 'field_description': 'Name'}),
+                (0, 0, {'name': 'x_length', 'ttype': 'float', 'field_description': 'Length'}),
+                (0, 0, {'name': 'x_color', 'ttype': 'integer', 'field_description': 'Color'}),
+                (0, 0, {'name': 'x_ripeness_id', 'ttype': 'many2one',
+                        'field_description': 'Ripeness','relation': 'x_banana_ripeness',
+                        'group_expand': True})
+            ]
+        })
+        cls.env['ir.default'].set(cls.test_model.model, 'x_ripeness_id', cls.ripeness_green)
+        # add a non-stored field, we'll check that it is rejected by the constraint
+        cls.non_stored_field = cls.env['ir.model.fields'].create({
+            'name': 'x_is_yellow',
+            'field_description': 'Is the banana yellow?',
+            'ttype': 'boolean',
+            'model_id': cls.test_model.id,
+            'store': False,
+            'depends': 'x_color',
+            'compute': "for banana in self:\n    banana['x_is_yellow'] = banana.x_color == 9"
+        })
+        cls.bananas = cls.env[cls.test_model.model].create([{
+            'x_name': 'Banana #1',
+            'x_length': 3.14159,
+            'x_color': 9,
+        }, {
+            'x_name': 'Banana #2',
+            'x_length': 0,
+            'x_color': 6,
+        }, {
+            'x_name': 'Banana #3',
+            'x_length': 10,
+            'x_color': 6,
+        }])
+        cls.addClassCleanup(cls.registry.leave_test_mode)
+        return res
+
+    def test_01_model_order_constraint(self):
+        """Check that the order constraint is properly enforced."""
+        VALID_ORDERS = ['id', 'id desc', 'id asc, x_length', 'x_color, x_length, create_uid']
+        for order in VALID_ORDERS:
+            self.test_model.order = order
+        INVALID_ORDERS = ['', 'x_wat', 'id esc', 'create_uid,', 'id, x_is_yellow']
+        for order in INVALID_ORDERS:
+            with self.assertRaises(ValidationError):
+                self.test_model.order = order
```

Comment:
This works by chance, because nothing is actually updated in database.
Using a savepoint would be cleaner (maybe `assertRaises` should do it systematically).

Reply:
Don't see why a savepoint is needed, `_validate_fields` is called in `write` and the last `create` call in the setup is sent to db directly. So I'm validating against updated values. What am I missing here?
Since I'm checking a python constraint, this should never make it to the database...

Reply:
The problem is: the python constraint crashes, but the value is actually in the database, and the update is actually not rolled back.  I know I am picky, but this is not 100% clean...

I propose this instead:
```python
for order in INVALID_ORDERS:
    with self.assertRaises(ValidationError), self.cr.savepoint():
        self.test_model.order = order
```
I am actually thinking about making `assertRaises` systematically use a savepoint guard, so that after this block, the database is just like before.

Reply:
I don't mind doing these changes and you being picky, i just wanted to understand what you meant (hence 'what am i missing' and not 'stop being a diva' 😜)

Indeed, you're right. Thx for the explanation :)